### DISCUSSION TITLE
Migration from 6.4.1 to 7.1.0 - removed classes/functions

### DISCUSSION THREAD

**Discussion by pawelryznar** [vladimir-bukhtoyarov/bucket4j#223] [Q&A] [answered]
Hi. I'm trying to migrate from 6.4.1 to 7.1.0. In my project I'm using:

```
    implementation("com.github.vladimir-bukhtoyarov:bucket4j-core:7.1.0")
    implementation("com.github.vladimir-bukhtoyarov:bucket4j-jcache:7.1.0")
```

I don't see how could I migrate that code, as it's deprecated (Bucket4j) or deleted (GridBucketState, JCache):
```
import io.github.bucket4j.Bucket4j
import io.github.bucket4j.grid.GridBucketState
import io.github.bucket4j.grid.jcache.JCache

    private val cache: Cache<String, GridBucketState> = cacheManager.getCache(rateLimitProperties.cacheName)
    private val buckets = Bucket4j.extension(JCache::class.java).proxyManagerForCache(cache)
.
.
.
val bucket = buckets.getProxy(principal.name, configuration)
val probe = bucket.tryConsumeAndReturnRemaining(1)
```

and in ehcache.xml config:
```
    <cache alias="rate-limit-buckets">
        <key-type>java.lang.String</key-type>
        <value-type>io.github.bucket4j.grid.GridBucketState</value-type>
        <expiry>
            <ttl>3600</ttl>
        </expiry>
        <heap>1000000</heap>
        <jsr107:mbeans enable-statistics="true"/>
    </cache>
```

**Comment by vladimir-bukhtoyarov** [comment]
@pawelryznar hello,

See [this](https://github.com/vladimir-bukhtoyarov/bucket4j-spring-boot-starter/commit/a53be83325d6e739b4661a5b88b5abae1e7015c5) as an example of migration. And do not forget to read the documentation https://bucket4j.com/7.1.0/toc.html#bucket4j-jcache

**Reply by pawelryznar** [reply]
Yes of course. I read it, but would be great if in the documentation would have some short upgrade guide, especially that version 7 introduces quite a lot of breaking changes.
It's hard to just look at some PR, and say why `GridBucketState` was used before, and now it's removed and why I should use `byte[]` instead.

**Reply by vladimir-bukhtoyarov** [reply]
Storing buckets as a flat array of bytes helps to be compatible with peer class loading. Modern JCache implementations allow to cross-node downloading of byte code of EntryProcessors, see for example the documentation https://ignite.apache.org/docs/latest/code-deployment/peer-class-loading 

But no one implementation provides cross-node classloading for cache values, for example by the hyperlink above you can see following note:
![image](https://user-images.githubusercontent.com/6761830/151801144-f5851b39-650c-4783-9403-43ab9958591d.png)
So storing state as bytes solves the limitations.


> I read it, but would be great if in the documentation would have some short upgrade guide, especially that version 7 introduces quite a lot of breaking changes.

Exactly, I want the same. But it is a non-profitable project that is supported in my free time, so it can not be ideal.

**Comment by pawelryznar** [comment, marked as answer]
I changed my code to the following:

```
import io.github.bucket4j.grid.jcache.JCacheProxyManager

    private val cache: Cache<String, ByteArray> = cacheManager.getCache(rateLimitProperties.cacheName)
    private val buckets = JCacheProxyManager(cache)
.
.
.
val bucket = buckets.builder().build(principal.name, configuration)
val probe = bucket.tryConsumeAndReturnRemaining(1)
```

and removed key-type and value-type from ehcache.xml:
```
    <cache alias="rate-limit-buckets">
        <expiry>
            <ttl>3600</ttl>
        </expiry>
        <heap>1000000</heap>
        <jsr107:mbeans enable-statistics="true"/>
    </cache>
```

It's working now, so I hope it's correct.

**Reply by vladimir-bukhtoyarov** [reply]
Yes. It is right way.

**Comment by gustavofariaas** [comment]
Hey @vladimir-bukhtoyarov , I've recently refactoring some code from java 11 to java 17, and I'm working with Ignite and bucket4J.

This is how my configuration is implemented:

`@Bean
    public ProxyManager<String> getProxyManager(IgniteCache<String, byte[]> cache) {
        return new IgniteProxyManager<>(cache);
    }

    @Bean
    public IgniteCache<String, byte[]> getCache() {
        IgniteConfiguration cfg = new IgniteConfiguration();

        cfg.setMetricsLogFrequency(0);
        cfg.setPeerClassLoadingEnabled(true);
        cfg.setDeploymentMode(DeploymentMode.SHARED);
        cfg.setClientMode(true);
        cfg.setPeerClassLoadingEnabled(true);

        TcpDiscoverySpi discoverySpi = new TcpDiscoverySpi();

        TcpDiscoveryVmIpFinder ipFinder = new TcpDiscoveryVmIpFinder();

        ipFinder.setAddresses(Collections.singletonList("127.0.0.1:7373"));

        discoverySpi.setIpFinder(ipFinder);
        cfg.setDiscoverySpi(discoverySpi);
        org.apache.ignite.Ignite ignite = Ignition.start(cfg);
        ignite.cluster().state(ClusterState.ACTIVE);
        IgniteCache<String, byte[]> cache = ignite.getOrCreateCache("buckets");
        return cache;
    }`

The issue I'm getting is when I call the function getProxyConfiguration which is returning to me an expecption:
javax.cache.processor.EntryProcessorException: class org.apache.ignite.binary.BinaryInvalidTypeException: io.github.bucket4j.grid.GridBucketState

Function: Optional<BucketConfiguration> buckConfig = proxyManager.getProxyConfiguration(bucketKey);

Even when GridBucketState is no longer available on the lib 8.0.1, I'm still getting this error, do you have some thoughts on this?

**Comment by vladimir-bukhtoyarov** [comment]
Hello @gustavofariaas

I have two ideas according to `BinaryInvalidTypeException`:
* You have not updated classpath on all nodes. Bucket4j jars should be present on every node in cluster.
* If classpath on every node updatedm but you still getting this error, then most likely it is not possible to deserialize previously persisted state of bucket, that was serialized in incompatible form. In such case I  would recommend to clear the cache, or use another cache.

**Reply by gustavofariaas** [reply]
thanks @vladimir-bukhtoyarov , I could resolve this issue by changing the cache to a new one, as the old was used by java 11 and GridBucketState, now it is working as expected.
